ratatui-widgets/src/logo.rs

```diff
@@ -168,6 +168,189 @@ impl Size {
+    "            ███ ██████████      ",
+    "          bb    █████████       ",
+    "         bxxb   █████████       ",
+    "        bxxxxb ██████████       ",
+    "       bxx█xxxb █████████       ",
+    "      bxx██xxxxb ████████       ",
+    "     bxxxxxxxxxxb ██████████    ",
+    "    bxxxxxxxxxxxxb ██████████   ",
+    "   bxxxxxxx██xxxxxb █████████   ",
+    "  bxxxxxxxxx██xxxxxb ████  ███  ",
+    " bxxxxxxxxxxxxxxxxxxb ██   ███  ",
+    "bxxxxxxxxxxxxxxxxxxxxb █   ███  ",
+    "bxxxxxxxxxxxxxxxxxxxxxb   ███   ",
+    " bxxxxxxxxxxxxxxxxxxxxxb ███    ",
+    "  bxxxxxxxxxxxxxxxxxxxxxb █     ",
+];
+
+/// A widget that renders the Ratatui mascot
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
+pub struct RatatuiMascot {
+    colors: MascotColors,
+}
+
+/// The colors used to render the Ratatui mascot
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+/// The colors used to render the Ratatui mascot
+pub struct MascotColors {
+    /// The color of the rat
+    pub rat: Color,
+    /// The color of the rat's eye
+    pub rat_eye: Color,
+    /// The color of the rat's hat
+    pub hat: Color,
+    /// The color of the terminal
+    pub term: Color,
+    /// The color of the terminal border
+    pub term_border: Color,
+}
+
+impl Default for MascotColors {
+    fn default() -> Self {
+        Self {
+            rat: Color::Rgb(204, 204, 204), // color of the website
+            hat: Color::Rgb(255, 255, 255),
+            rat_eye: Color::Rgb(8, 8, 8),
+            term: Color::Rgb(8, 8, 8),
+            term_border: Color::Rgb(188, 188, 188),
+        }
+    }
+}
+
+use itertools::Itertools; // tuples();
+
+impl RatatuiMascot {
+    /// Create a new Ratatui mascot widget
+    pub const fn new(colors: MascotColors) -> Self {
+        Self { colors }
+    }
+
+    /// Tiny help to set the color of the rat's eye
+    #[must_use]
+    pub const fn set_eye_color(self, rat_eye: Color) -> Self {
+        Self {
+            colors: MascotColors {
+                rat_eye,
+                ..self.colors
+            },
+        }
+    }
+
+    /// Use half block characters to render a logo based on the `RATATUI_LOGO` const.
+    ///
+    /// The logo is rendered in three colors, one for the rat, one for the terminal, and one for the
+    /// rat's eye. The eye color alternates between two colors based on the selected row.
+    pub fn render_mascot(&self, area: Rect, buf: &mut Buffer) {
+        let rat_color = self.colors.rat;
+        let term_color = self.colors.term;
+        let hat_color = self.colors.hat;
+        let border_color = self.colors.term_border;
+        for (y, (line1, line2)) in RATATUI_MASCOT.iter().tuples().enumerate() {
+            for (x, (ch1, ch2)) in line1.chars().zip(line2.chars()).enumerate() {
+                let x = area.left() + x as u16;
+                let y = area.top() + y as u16;
+                let cell = &mut buf[(x, y)];
+                match (ch1, ch2) {
+                    ('█', '█') => {
```

Comment:
This logic was fine for 2 colors, but once you go beyond that it's worth instead mapping the characters a bit differently. I might be wrong on this as the (literal) edge cases here seem to be a bit hairy, but I think it culd be as simple as something like the following (untested code):
```rust

// given two cells which make up the top and bottom of the character, 
// Foreground color should be the non-space, non-terminal
let (fg, bg) = match (ch1, ch2) {
    (' ', ' ') => (Color::default(), Color::default()),
    (c, ' ') => (color_for(c), Color::default()),
    (' ', c) => (color_for(c), Color::default()),
    ('x', c) = (color_for(c), TERM_BG), // treat the terminal background cells as empty
    (c, 'x') = (TERM_BG, color_for(c)), // rather than having a character
    (c1, c2) => (color_for(c1), color_for(c2)),
}
// symbol should make the empty space or terminal bg as the empty part of the block
let symbol = match (ch1, ch2) {
    (' ', ' ') => ' ',
    ('x', 'x') => ' ', // treat the terminal background as an empty cell
    (c, ' ' | 'x') => '▀',
    (' ' | 'x', c) => '▄',
    (c, d) if c == d => '█',
    (c, d) => '▀',
}
```

Reply:
This broke the cursor and border pixels - it's likely one of those difficult to get exactly right bits (edge cases :D).

Original
<img width="597" alt="image" src="https://github.com/user-attachments/assets/d6942431-87eb-45b1-9a73-c5555a36f386" />

First cut:
![image](https://github.com/user-attachments/assets/7fa502ee-9735-4475-8dc5-3ac937e59cbb)

Current:
![image](https://github.com/user-attachments/assets/5b5df8c3-8b96-42fc-ad42-ebb5128f7db3)

The terminal border pixels and cursor are now broken.

Reply:
![image](https://github.com/user-attachments/assets/751e939c-f229-45e3-b429-08fa530d5809)

Is it correct now ?